app/controllers/pensions/annualAllowance/PensionSchemeTaxReferenceController.scala

```diff
@@ -0,0 +1,111 @@
+                                                    pensionSchemeTaxReferenceView: PensionSchemeTaxReferenceView,
+                                                    appConfig: AppConfig,
+                                                    pensionSessionService: PensionSessionService,
+                                                    errorHandler: ErrorHandler,
+                                                    clock: Clock) extends FrontendController(cc) with I18nSupport {
+
+
+  private def prefillValue(pstrListOpt: Option[Seq[String]], pensionSchemeTaxReference: Option[Int], user: User): Form[String] = {
+    val emptyForm: Form[String] = PensionSchemeTaxReferenceForm.pensionSchemeTaxReferenceForm(user.isAgent)
+    (pstrListOpt, pensionSchemeTaxReference) match {
+      case (Some(pstrList), Some(pstrIndex)) => if (pstrList.size < pstrIndex) emptyForm else emptyForm.fill(pstrList(pstrIndex))
+      case (None, Some(_)) =>
+        //Person has request a certain pstr but it doesn't exist on the list, this shouldn't happen unless it's through an API request
+        //in this case the prefill value will be None.
+        emptyForm
+      case (Some(_), None) => emptyForm
+      case (None, None) => emptyForm
+    }
+  }
+
+  def show(taxYear: Int, pensionSchemeTaxReference: Option[Int]): Action[AnyContent] = authAction.async { implicit request =>
+
+    pensionSessionService.getPensionsSessionDataResult(taxYear, request.user) {
+      case Some(data) =>
+        val form = prefillValue(
+          data.pensions.pensionsAnnualAllowances.pensionSchemeTaxReference,
+          pensionSchemeTaxReference,
+          request.user
+        )
+        Future.successful(Ok(pensionSchemeTaxReferenceView(form, taxYear, pensionSchemeTaxReference)))
+      case _ =>
+        Future.successful(Redirect(PensionsSummaryController.show(taxYear)))
+    }
+  }
+
+  def submit(taxYear: Int, pensionSchemeTaxReference: Option[Int]): Action[AnyContent] = authAction.async {
+    implicit request =>
+      PensionSchemeTaxReferenceForm.pensionSchemeTaxReferenceForm(request.user.isAgent).bindFromRequest().fold(
+        formWithErrors => Future.successful(BadRequest(pensionSchemeTaxReferenceView(formWithErrors, taxYear, pensionSchemeTaxReference))),
+        pensionScheme => {
+          pensionSessionService.getPensionsSessionDataResult(taxYear, request.user) {
+            data =>
+              val pensionsCYAModel: PensionsCYAModel = data.map(_.pensions).getOrElse(PensionsCYAModel.emptyModels)
+              val viewModel = pensionsCYAModel.pensionsAnnualAllowances
+              val newPensionSchemeTaxRef = Seq(pensionScheme)
+
+              val updatedPstrList: Seq[String] = (viewModel.pensionSchemeTaxReference, pensionSchemeTaxReference) match {
+                case (Some(pstrList), Some(pstrIndex)) =>
+                  //not sure what the action should be here if a user attempts to update an out of bound index ?
```

Comment:
Yes good point and need to consider this as you suggest.
If the index is found to be inconsistent with session we should reject the request.
This might happen if they bookmarked a URL, if they typed it in the web page URL and was remembered, or maybe if they had 2 screens and removed one on one screen and then tried to update on another..

Several ways it might happen but we should not allow the update/create where the data is not consistent.

The best thing in this situation would be to direct them to the PSTR summary page so they have to pick a valid PSTR (or the summary page for now with a TODO as it does not exist yet).